dev/src/bulk-writer.ts

```diff
@@ -738,29 +851,85 @@ export class BulkWriter {
    * committed, sends the next group of ready batches.
    *
+   * @param batchCompletedDeferred A deferred promise that resolves when the
+   * batch has been sent and received.
    * @private
    */
-  private sendBatch(batch: BulkCommitBatch): void {
-    const success = this.rateLimiter.tryMakeRequest(batch.opCount);
+  private sendBatch(
+    batch: BulkCommitBatch,
+    batchQueue: BulkCommitBatch[],
+    batchCompletedDeferred: Deferred<void>
+  ): Promise<void> {
+    const success = this._rateLimiter.tryMakeRequest(batch.opCount);
     assert(success, 'Batch should be under rate limit to be sent.');
-    batch.bulkCommit().then(() => {
+    return batch.bulkCommit().then(() => {
       // Remove the batch from the BatchQueue after it has been processed.
-      const batchIndex = this.batchQueue.indexOf(batch);
+      const batchIndex = batchQueue.indexOf(batch);
       assert(batchIndex !== -1, 'The batch should be in the BatchQueue');
-      this.batchQueue.splice(batchIndex, 1);
+      batchQueue.splice(batchIndex, 1);
 
-      this.sendReadyBatches();
+      batchCompletedDeferred.resolve();
+      this._pendingBatches.delete(batchCompletedDeferred.promise);
+
+      this.sendReadyBatches(batchQueue);
     });
   }
 
+  /**
+   * Schedules and runs the provided operation.
+   */
+  private async _executeWrite(
+    documentRef: firestore.DocumentReference,
+    operationType: 'create' | 'set' | 'update' | 'delete',
+    operationFn: (bulkCommitBatch: BulkCommitBatch) => Promise<WriteResult>
+  ): Promise<WriteResult> {
+    // A deferred promise that resolves when operationFn completes.
+    const operationCompletedDeferred = new Deferred<void>();
+    this._pendingOps.add(operationCompletedDeferred.promise);
+    try {
+      for (let failedAttempts = 0; ; ++failedAttempts) {
+        const batchQueue =
+          failedAttempts > 0 ? this._retryBatchQueue : this._batchQueue;
+        const bulkCommitBatch = this.getEligibleBatch(batchQueue);
+
+        // Send ready batches if this is the first attempt. Subsequent retry
+        // batches are scheduled after the initial batch returns.
+        if (failedAttempts === 0) {
+          this.sendReadyBatches(batchQueue);
+        }
+
+        try {
+          const operationResult = await operationFn(bulkCommitBatch);
+          this._successFn(documentRef, operationResult);
+          return operationResult;
+        } catch (error) {
+          const bulkWriterError = new BulkWriterError(
+            error.code,
+            error.message,
+            documentRef,
+            operationType,
+            failedAttempts
+          );
+          const shouldRetry = this._errorFn(bulkWriterError);
+          if (!shouldRetry) {
+            throw bulkWriterError;
+          }
+        }
+      }
+    } finally {
+      operationCompletedDeferred.resolve();
+      this._pendingOps.delete(operationCompletedDeferred.promise);
+    }
```

Comment:
I think the double try/catch block is a bit harder to parse then just calling these two lines twice.